File: agent_workflow/project_doc.py
"""Read-only project-doc lookup, listing, staleness and contract checks."""
from __future__ import annotations
import argparse,json,re,subprocess
from pathlib import Path
from .protocol import write_json
def metadata(text):
    m=re.search(r"(?ms)^---\n(.*?)\n---\n?(.*)$",text.replace("\r\n","\n"));
    if not m:return None
    data={}
    for line in m.group(1).splitlines():
        x=re.match(r"^([a-z_]+):\s*(.*)$",line)
        if x:
            value=x.group(2).strip()
            try:value=json.loads(value)
            except Exception:pass
            data[x.group(1)]=value
    if not data.get("doc_type"):return None
    data["body"]=m.group(2);return data
def git(root,args):return subprocess.run(["git","-C",str(root),*args],stdout=subprocess.PIPE,stderr=subprocess.DEVNULL,check=False)
def doc_list(root,doc_root):
    base=(Path(root).resolve()/doc_root).resolve(); out=[]
    if not base.is_dir():return out
    for path in base.rglob("*.md"):
        data=metadata(path.read_text(encoding="utf-8-sig"));
        if not data:continue
        rel=str(path.relative_to(Path(root).resolve())).replace("\\","/"); covers=data.get("covers",[]); covers=[covers] if isinstance(covers,str) else [str(x) for x in covers];
        def stamp(args):
            p=git(root,args); value=p.stdout.decode().strip()
            try:return int(value)
            except ValueError:return None
        # An empty covers list (architecture/dataflow/glossary) must never fall through to
        # a bare `-- ` pathspec: git treats that as "no restriction" and reports the whole
        # repo's latest activity, which would make every such doc spuriously stale the
        # moment anything else in the repo changes.
        doc_time=stamp(["log","-1","--format=%ct","--",rel])
        code_time=stamp(["log","-1","--format=%ct","--",*covers]) if covers else None
        dirty=bool(git(root,["status","--porcelain","--",*covers]).stdout.strip()) if covers else False
        out.append({"path":str(path),"doc_type":data.get("doc_type"),"covers":covers,"stale":doc_time is not None and code_time is not None and code_time>doc_time,"stale_pending":dirty and not bool(git(root,["status","--porcelain","--",rel]).stdout.strip()),"body":data.get("body","")})
    return out
def run(action,doc="",repo_root=".",paths=None,doc_root="docs"):
    root=Path(repo_root).resolve(); items=doc_list(root,doc_root); key=action.casefold()
    if key=="check":
        target=Path(doc); issues=[]
        if not target.is_file():return [{"path":str(target),"issues":["document does not exist"]}]
        data=metadata(target.read_text(encoding="utf-8-sig")); allowed={"architecture","structure","dataflow","flow","module","api","decision","glossary"}
        if not data:issues.append("missing or invalid frontmatter")
        else:
            if data.get("doc_type") not in allowed:issues.append(f"unknown doc_type: {data.get('doc_type')}")
            covers=data.get("covers",[]); covers=[covers] if isinstance(covers,str) else covers
            if data.get("doc_type") not in {"architecture","structure","dataflow","glossary"} and not covers:issues.append("covers must not be empty")
            required={"structure":["Layout","Placement rules","Unverified"],"flow":["Trigger","Steps","Failure modes","Unverified"],"module":["Responsibility","Entrypoints","Flow","Shared state","Invariants and gotchas","Unverified"],"api":["Endpoint","Auth","Request","Response","Errors","Invariants and gotchas","Unverified"],"decision":["Context","Decision","Alternatives","Consequences"],"glossary":["Terms"]}.get(data.get("doc_type"),[])
            for name in required:
                m=re.search(rf"(?ms)^## {re.escape(name)}\s*\n(.*?)(?=^## |\Z)",data.get("body",""));
                if not m or not m.group(1).strip() or re.match(r"^<.*>$",m.group(1).strip()):issues.append(f"missing section: {name}")
        return [{"path":str(target),"issues":issues}]
    results=[]
    for item in items:results.append({k:v for k,v in item.items() if k!="body"}|{"matched_by":[]})
    if key=="list":return results
    if key=="stale":return [x for x in results if x["stale"] or x["stale_pending"]]
    queries=[]
    for value in paths or []:
        target=(root/value).resolve() if not Path(value).is_absolute() else Path(value).resolve()
        try:queries.append(str(target.relative_to(root)).replace("\\","/").casefold())
        except ValueError:raise RuntimeError(f"path is outside the repo root: {value}")
    matched=set()
    for result,item in zip(results,items):
        for query in queries:
            if any(query==entry.casefold() or (entry.endswith("/") and (query.startswith(entry.casefold()) or query==entry.rstrip("/").casefold())) for entry in item["covers"]):result["matched_by"].append(query);matched.add(query)
        result["matched_by"]=sorted(set(result["matched_by"]))
    results.sort(key=lambda x:max([len(y) for y in x["matched_by"]] or [0]),reverse=True);return {"docs":results,"uncovered":[x for x in queries if x not in matched]}
def main(argv=None):
    p=argparse.ArgumentParser();p.add_argument("--action",required=True,choices=("Lookup","List","Stale","Check"));p.add_argument("--doc",default="");p.add_argument("--repo-root",default=".");p.add_argument("--doc-root",default="docs");p.add_argument("--paths",nargs="*");a=p.parse_args(argv);write_json(run(a.action,a.doc,a.repo_root,a.paths,a.doc_root));return 0
if __name__=="__main__":raise SystemExit(main())

File: src/project-doc.ts
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { join, relative, resolve } from "node:path";
import { JsonObject, git, isWithin, output, readJson, sha256 } from "./core.js";
import { taskWrite } from "./lifecycle/transitions.js";
type Options = Map<string, string | boolean | string[]>;
const text = (o: Options, n: string, f = "") => typeof o.get(n) === "string" ? o.get(n) as string : f;
const paths = (o: Options) => text(o, "paths").split(",").map((v) => v.trim()).filter(Boolean);
// Repo-wide overviews are candidates, not implicit reads. Returning their digest lets an agent
// reuse a document already read in the same session without reopening every overview on each lookup.
const OVERVIEW_TYPES = ["architecture", "structure", "dataflow", "glossary"];
const DOC_TYPES = new Set(["architecture", "structure", "dataflow", "flow", "module", "api", "decision", "glossary"]);
const REQUIRED_SECTIONS: Record<string, string[]> = {
  structure: ["layout", "placement rules", "unverified"],
  flow: ["trigger", "steps", "failure modes", "unverified"],
  module: ["responsibility", "entrypoints", "flow", "shared state", "invariants and gotchas", "unverified"],
  api: ["endpoint", "auth", "request", "response", "errors", "invariants and gotchas", "unverified"],
  decision: ["context", "decision", "alternatives", "consequences"],
  glossary: ["terms"]
};
// Both YAML sequence styles are accepted. Reading only the block form made a flow-style
// `covers: ["a", "b"]` parse as an empty list, which does not fail anywhere: the document simply
// stops matching any path and drops out of every Lookup without a single error to notice.
function coversOf(header: string): string[] {
  const flow = header.match(/^covers:[ \t]*\[([^\]]*)\]/m);
  const items = flow ? flow[1].split(",")
    : (header.match(/^covers:\s*\r?\n((?:\s*-.*\r?\n?)*)/m)?.[1] || "").split(/\r?\n/).map((line) => line.replace(/^\s*-\s*/, ""));
  return items.map((item) => item.trim().replace(/^(["'])(.*)\1$/, "$2").trim()).filter(Boolean);
}
function docs(root: string, directory: string): JsonObject[] {
  const base = join(root, directory); if (!existsSync(base)) return [];
  const found: JsonObject[] = [];
  const walk = (path: string): void => {
    for (const entry of readdirSync(path, { withFileTypes: true })) {
      const next = join(path, entry.name);
      if (entry.isDirectory()) walk(next);
      else if (entry.isFile() && next.endsWith(".md")) {
        const body = readFileSync(next, "utf8");
        const header = body.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n?/)?.[1];
        const type = header?.match(/^doc_type:\s*(.+)$/m)?.[1].trim().replace(/^['"]|['"]$/g, "");
        if (!type) continue;
        found.push({ path: next, doc_type: type, covers: coversOf(header || ""), content_sha256: sha256(body) });
      }
    }
  };
  walk(base); return found;
}
function checkDocument(root: string, docRoot: string, path: string): string[] {
  if (!existsSync(path)) return ["document does not exist"];
  if (!isWithin(path, join(root, docRoot))) return ["document is outside the configured doc-root"];
  const body = readFileSync(path, "utf8");
  const headerMatch = body.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n?/);
  if (!headerMatch) return ["missing or invalid frontmatter"];
  const header = headerMatch[1];
  const rawType = header.match(/^doc_type:\s*(.+)$/m)?.[1].trim().replace(/^['"]|['"]$/g, "") || "";
  const issues: string[] = [];
  if (!DOC_TYPES.has(rawType)) issues.push(`doc_type must be one of: ${[...DOC_TYPES].join(", ")}`);
  const coverValues = coversOf(header);
  if (["flow", "module", "api", "decision"].includes(rawType) && !coverValues.length) issues.push(`${rawType} documents require a non-empty covers list`);
  for (const cover of coverValues) {
    if (!cover || cover.includes("..") || cover.includes("[") || cover.includes("]") || cover.includes("\\") || cover.startsWith("/")) issues.push(`invalid covers path: ${cover}`);
  }
  const headings = new Set([...body.matchAll(/^##\s+(.+?)\s*$/gm)].map((match) => match[1].trim().toLowerCase()));
  for (const required of REQUIRED_SECTIONS[rawType] || []) if (!headings.has(required)) issues.push(`missing required section: ## ${required}`);
  if (OVERVIEW_TYPES.includes(rawType)) {
    const siblings = docs(root, docRoot).filter((item) => String(item.doc_type) === rawType && String(item.path) !== path);
    if (siblings.length) issues.push(`duplicate singleton doc_type '${rawType}': ${siblings.map((item) => String(item.path)).join(", ")}`);
  }
  return issues;
}

function taskStatePath(value: string): string {
  const candidate = resolve(value);
  return candidate.toLowerCase().endsWith(".json") ? candidate : join(candidate, "task.json");
}

function projectDocPath(root: string, path: string): string {
  return relative(root, path).replaceAll("\\", "/");
}

type ProjectDocState = { read: Set<string>; digests: Map<string, string> };
function projectDocState(root: string, taskValue: string): ProjectDocState {
  if (!taskValue) return { read: new Set(), digests: new Map() };
  const state = readJson(taskStatePath(taskValue)) as JsonObject;
  const project = state.project_docs && typeof state.project_docs === "object" ? state.project_docs as JsonObject : {};
  const read = new Set((Array.isArray(project.read) ? project.read : []).map((value) => String(value).replaceAll("\\", "/")));
  const digests = new Map<string, string>();
  for (const item of Array.isArray(project.digests) ? project.digests : []) {
    if (!item || typeof item !== "object") continue;
    const entry = item as JsonObject;
    if (typeof entry.path === "string" && typeof entry.content_sha256 === "string") digests.set(entry.path.replaceAll("\\", "/"), entry.content_sha256);
  }
  return { read, digests };
}

function digestStatus(root: string, item: JsonObject, state: ProjectDocState, taskValue: string): string {
  if (!taskValue) return "untracked";
  const path = projectDocPath(root, String(item.path));
  if (!state.read.has(path)) return "unread";
  const recorded = state.digests.get(path);
  if (!recorded) return "digest_missing";
  return recorded === String(item.content_sha256) ? "reusable" : "stale";
}

function remember(options: Options): number {
  const taskValue = text(options, "task-path");
  if (!taskValue) throw new Error("project-doc Remember requires --task-path");
  const root = resolve(text(options, "repo-root", process.cwd()));
  const docRoot = text(options, "doc-root", "docs");
  const requested = paths(options);
  if (!requested.length) throw new Error("project-doc Remember requires --paths");
  const entries = docs(root, docRoot);
  const byPath = new Map(entries.map((item) => [projectDocPath(root, String(item.path)), item]));
  const remembered: JsonObject[] = [];
  const issues: string[] = [];
  for (const requestedPath of requested) {
    const absolute = resolve(root, requestedPath);
    const canonical = projectDocPath(root, absolute);
    const item = byPath.get(canonical);
    if (!item) { issues.push(`${requestedPath}: document is missing or has invalid frontmatter`); continue; }
    const documentIssues = checkDocument(root, docRoot, absolute);
    if (documentIssues.length) { issues.push(`${requestedPath}: ${documentIssues.join("; ")}`); continue; }
    remembered.push({ path: canonical, content_sha256: String(item.content_sha256) });
  }
  if (issues.length) { output({ valid: false, errors: issues }); return 1; }
  const state = readJson(taskStatePath(taskValue)) as JsonObject;
  const current = state.project_docs && typeof state.project_docs === "object" ? state.project_docs as JsonObject : {};
  const read = new Set((Array.isArray(current.read) ? current.read : []).map((value) => String(value).replaceAll("\\", "/")));
  const digestMap = new Map<string, string>();
  for (const item of Array.isArray(current.digests) ? current.digests : []) {
    if (!item || typeof item !== "object") continue;
    const entry = item as JsonObject;
    if (typeof entry.path === "string" && typeof entry.content_sha256 === "string") digestMap.set(entry.path.replaceAll("\\", "/"), entry.content_sha256);
  }
  for (const item of remembered) { read.add(String(item.path)); digestMap.set(String(item.path), String(item.content_sha256)); }
  const nextProject = {
    read: [...read].sort(),
    updated: (Array.isArray(current.updated) ? current.updated : []).map(String).sort(),
    digests: [...digestMap.entries()].sort(([left], [right]) => left.localeCompare(right)).map(([path, content_sha256]) => ({ path, content_sha256 }))
  };
  const previous = JSON.stringify(current); const next = JSON.stringify(nextProject);
  let stateRevision = Number(state.state_revision || 0);
  if (previous !== next) {
    const code = taskWrite(taskValue, { project_docs: nextProject }, undefined, root, false, undefined, false);
    if (code !== 0) { output({ valid: false, errors: ["project-doc Remember could not update task state"] }); return code; }
    stateRevision = Number((readJson(taskStatePath(taskValue)) as JsonObject).state_revision || stateRevision);
  }
  output({ valid: true, task: taskStatePath(taskValue), state_revision: stateRevision, remembered });
  return 0;
}

export function projectDoc(options: Options): number {
  const action = text(options, "action"); const root = resolve(text(options, "repo-root", process.cwd())); const docRoot = text(options, "doc-root", "docs"); const result = docs(root, docRoot);
  if (action === "List") { output(result); return 0; }
  if (action === "Check") { const requested = text(options, "doc"); const path = resolve(root, requested); const issues = checkDocument(root, docRoot, path); output([{ path, issues }]); return issues.length ? 1 : 0; }
  if (action === "Stale") { output(result.filter((item) => { const rel = relative(root, String(item.path)); const covered = Array.isArray(item.covers) ? item.covers.map(String) : []; const docTime = Number(git(root, ["log", "-1", "--format=%ct", "--", rel]).stdout || 0); const codeTime = covered.length ? Number(git(root, ["log", "-1", "--format=%ct", "--", ...covered]).stdout || 0) : 0; return codeTime > docTime; })); return 0; }
  if (action === "Remember") return remember(options);
  if (action !== "Lookup") throw new Error(`unsupported project-doc action: ${action}`);
  const taskValue = text(options, "task-path"); const prior = projectDocState(root, taskValue);
  const requested = paths(options).map((path) => path.replaceAll("\\", "/").toLowerCase()); const matched = new Set<string>();
  const mapped = result.map((item) => {
    const covers = Array.isArray(item.covers) ? item.covers.map(String) : [];
    const matchedBy = requested.filter((path) => covers.some((cover) => path === cover.toLowerCase() || (cover.endsWith("/") && path.startsWith(cover.toLowerCase()))));
    matchedBy.forEach((path) => matched.add(path));
    return { ...item, doc_type: String(item.doc_type), matched_by: matchedBy, digest_status: digestStatus(root, item, prior, taskValue) };
  }).filter((item) => !OVERVIEW_TYPES.includes(String(item.doc_type)) && item.matched_by.length > 0).sort((left, right) => Number(right.matched_by.length) - Number(left.matched_by.length));
  const overview_candidates = result.filter((item) => OVERVIEW_TYPES.includes(String(item.doc_type))).map((item) => ({ path: item.path, doc_type: item.doc_type, content_sha256: item.content_sha256, reason: "repo-wide overview; read when the compiled exploration profile or task impact requires it", digest_status: digestStatus(root, item, prior, taskValue) }));
  output({ docs: mapped, overview_candidates, uncovered: requested.filter((path) => !matched.has(path)) }); return 0;
}
